<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">

    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-dark.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">

    <title>Dismiss by uhkniazi</title>
  </head>

  <body>

    <header>
      <div class="container">
        <h1>Dismiss</h1>
        <h2>DISMISS is an R script, which as an additional step in MeDIP-Seq data analysis workflow, enables the allocation of strands to methylated DNA regions. It does this by analyzing the proportions of first mate reads aligning to the methylated locus from the plus and minus strands.</h2>

        <section id="downloads">
          <a href="https://github.com/uhkniazi/dismiss/zipball/master" class="btn">Download as .zip</a>
          <a href="https://github.com/uhkniazi/dismiss/tarball/master" class="btn">Download as .tar.gz</a>
          <a href="https://github.com/uhkniazi/dismiss" class="btn btn-github"><span class="icon"></span>View on GitHub</a>
        </section>
      </div>
    </header>

    <div class="container">
      <section id="main_content">
        <h1><a id="dismiss" class="anchor" href="#dismiss" aria-hidden="true"><span class="octicon octicon-link"></span></a>dismiss</h1>

        <h2>Quick Reference</h2>
        <p>Dismiss is a collection of functions implemented in R  and require the following to be present on the system:</p>
          <ul>
            <li>I.	R (version 3.1.2)</li>
            <li>II.	Bioconductor ,  (version 3.0) with packages</li>
            <ul>
              <li>A.	GenomicAlignments  </li>
              <li>B.	GenomicFeatures  </li>
              <li>C.	rtracklayer  </li>
            </ul>
          </ul>

        <p>To run dismiss, the following files should be in the current path or directory:</p>
        <ul>
          <li>I.	dismiss_header.R (the script with all the required functions)</li>
          <li>II.	dismiss_macs2_extractor.R (utility script to process MACS2  results)</li>
          <li>III.	Tab separated xls results file for methylated regions for peak calling performed by MACS2. </li>
          <li>IV.	Alignment file in .bam format (the same file used to perform peak calling)</li>
          <li>V.	Bam index file with .bai extension (if bam file is called file.bam then index file should be called file.bam.bai)</li>
        </ul>
        
        <p>In order to simplify the process of strand separation, a utility script dismiss_macs2_extractor.R has been provided to perform strand separation on peak calling performed on MeDIP-Seq alignment data using MACS2. The script can be run from the command line, provided the above mentioned conditions are met.</p>

        <code>Rscript dismiss_macs2_extractor.R MACS2_result.xls File.bam [p or s]</code>

        <p>where the last argument is essential and should be set to either</p>
        <ul>
          <li>p = for paired end data OR</li>
          <li>s = for single end data</li>
        </ul>
        
        <p>The functions provided with the script dismiss_header.R can be used directly in oneâ€™s own R scripts, and strand separation performed on ranged data of GRanges object type. This ranged data would have been produced by a MeDIP-Seq analysis software of choice e.g. MEDIPS , MACS2 etc. (see section 2.3 for description of the function to call)</p>

        <h2>2.0 General Information</h2>
        <h3>2.1 What is DISMISS</h3>
        <p>DISMISS, which as an additional step in MeDIP-Seq data analysis workflow, enables the allocation of  strands to methylated DNA regions. It does this by analyzing analysing the proportions of of first mate read aligning to the methylated locus from the plus and minus strands.</p>

        <h3>2.2 Installation notes</h3>
        <p>DISMISS uses the base R package and Bioconductor libraries. R (version 3.1.2) can be installed by visiting http://cran.r-project.org/. Once R has been installed, then Bioconductor (version 3.0) can be installed directly (http://www.bioconductor.org/install/) through R by typing:</p>
        <code>
        source("http://bioconductor.org/biocLite.R")
        biocLite()
        </code>
        <p>The bioconductor packages GenomicAlignments, GenomicFeatures and rtracklayer can each be installed after these steps by typing e.g.</p>
        <code>biocLite("GenomicAlignments") </code>
        <p></p>
        <h3>2.3 How does it work</h3>
        <p>DISMISS can be used in two ways: 1) a more simplified script that processes results from MACS2 and outputs results as an R object and gff3 files - using dismiss_macs2_extractor.R; 2) a more involved method where the functions can be called directly in R. </p>

        <h3>2.3.1 Simple method.</h3>
        <p>This method involves a call to the script dimiss_macs2_extractor.R which performs the following steps:</p>
        <ol>
          <li>Process command line arguments by setting variables</li>
          <ol>
            <li>macs = name of the MACS2 results excel file.</li>
            <li>bam = name of the MeDIP-Seq alignment file in bam format</li>
            <li>paired = TRUE; if the command line argument was p or else s</li>
          </ol>
          <li>Source the main script with all functions dismiss_header.R</li>
          <li></li>Set the Genomic Ranges object oGRdismiss</li>
          <ol>
            <li>Call function f_oGRMACS2toGRanges with argument macs, to convert tab separated ranged data into a GRanges object oGRdismiss</li>
          </ol>
          <li>Separate the strands by calling the function f_oGRSeparateStrands with arguments oGRdismiss, bam and paired variables.</li>
          <li>Save the output as three gff3 files and an R object containing oGRdismiss object.</li>
        </ol>
        
        <h3>2.3.2 Involved method.</h3>
        <p>This is recommended for users who are more familiar with R, and would like to produce Genomic Ranges objects themselves. It is assumed that these Genomic Ranges objects are ranges over the genome where methylated signal is present. This method involves the use of the function f_oGRSeparateStrands. Each of the functions used here are described in more detail later in section 3.</p>

        <ul>
          <li>1.	Call the function with arguments</li>
          <ul>
            <li>1.1.	oGRsignal.strand = a GRanges object that has the strand value set to * (i.e. double stranded).</li>
            <li>1.2.	csFile = a character string with the name of the bam file e.g. infile.bam</li>
            <ul>
              <li>1.2.1.	the function assumes that there is a bam index file infile.bam.bai in the same directory</li>
            </ul>
            <li>1.3.	bPaired = a boolean (TRUE/FALSE) variable - if the data is paired end or single end.</li>
          </ul> 
          <li>2.	Count the first mate reads aligning to each range/region in the GRanges object, in a matrix called mFirstMate. <code>mFirstMate = f_mFastCountMateReadsOverRanges(oGRsignal.strand, bam.file = csFile, bPaired)</code></li>
          <li>3.	Using the mFirstMate matrix, apply the function over each row of the matrix, to check if the first mate reads from the plus (mp) and minus (mm) strands are equal under a poisson likelihood.<code>t = apply(mFirstMate[,c('mp', 'mm')], 1, f_bComparePoissonLiklihoodApply)</code></li>
          <li>4.	The function f_bComparePoissonLiklihoodApply will fit a likelihood function on the two values 'mp', 'mm' for each row in mFirstMate.</li>
          <li>5.	The variable t returned by f_bComparePoissonLiklihoodApply is an array of boolean values TRUE/FALSE of length equal to the number of rows in mFirstMate, with a TRUE if the two read counts on the plus and minus sides are equal or else:</li>
          <ul>
            <li>5.1.	Assign the peak range to plus if mFirstMate[i,'mp'] > mFirstMate[i,'mm'].</li>
            <li>5.2.	Else assign to minus strand.</li>
          </ul>
          <li>6.	Set the strand value for each GRanges index to *, + or - depending on the results from step 5.</li>
          <li>7.	Add some additional information to the mcols data of oGRsignal.strand</li>
          <ul>
            <li>7.1.	strand_fac = 2 (both), 1 (plus), 0 (minus) - factor variable to represent strands</li>
            <li>7.2.	mp = first mate reads from plus strand</li>
            <li>7.3.	mm = first mate reads from minus strand</li>
            <li>7.4.	bp = both plus, reads from first and second mate on plus strand</li>
            <li>7.5.	bm = both minus, reads from first and second mate on minus strand</li>
            <ul>
              <li>7.5.1.	bp and bm are the same as mp and mm in case of single end reads</li>
            </ul>
            <li>7.6.	p.plus = binomial probability of fraction of reads from mp</li>
            <li>7.7.	p.minus = binomial probability of fraction of reads from mm</li>
            <ul>
              <li>7.7.1.	calculated as:</li>
              <code>mp = mFirstMate[,'mp'] + 1 # add a one to avoid divisions by zero</code><br>
              <code>mm = mFirstMate[,'mm'] + 1</code><br>
              <code>n = mp + mm</code><br>
              <code>p.plus = mp/n</code><br>
              <code>p.minus = 1-p.plus</code><br>
            </ul>
          </ul>  
          <li>8.	Other variables associated with mcols data are values already included in the oGRsignal.strand data structure by the user.</li>
        </ul>  
      </section>
    </div>

    
  </body>
</html>
